### PR TITLE
Fixed: output directory generation not working as expected when --output is passed a directory name instead of filename

### LINKED ISSUE #39: [BugReport] output directory generation not working as expected.
https://github.com/moest-np/center-randomize/blob/607f3db84e5e589f49cf8c9cb165dfa24c0e0b32/school_center.py#L163C44-L163C53

if your comandline argument has ouptput unter a directory structure. thsi produces Directory not found Error.

Steps to reproduce:
python3 school_center.py sample_data/schools_grade12_2081.tsv sample_data/centers_grade12_2081.tsv sample_data/prefs.tsv -o sample_data/output_main.tsv
Traceback (most recent call last):
  File "/Users/user/Documents/MOEST/center-randomize/school_center.py", line 165, in <module>
    open(OUTPUT_DIR + args.output, 'w+', encoding='utf-8') as a_file:
FileNotFoundError: [Errno 2] No such file or directory: 'results/sample_data/output_main.tsv'

### PR DESCRIPTION
**Pull request by sebakthapa**
Fixes #39 

## The solved problem:
 - when `-o` or `--output` was not passed or when `filename` was passed from CLI, it was working well
 -  But when passed directory name, it was throwing error

**It is solved now**

## Changes done
The fixed code now uses `os.path` module to extract `dirname` and `filename` from the provided args and uses them accordingly.

### CODE REVIEW

**Comment by hariacharya80** [comment]
Please use a closing statement to close the linked issue automatically after merging, try adding
`closes #IssueNumber`

**Review by hariacharya80 — APPROVED** [review]
LGTM!

File: school_center.py [hunk 1]
```diff
@@ -205,10 +198,24 @@ def is_allocated(scode1: str, scode2: str) -> bool:
...
+def get_output_filename():
+    return path.basename(args.output)
+
+
+output_dirname = get_output_dir()
+output_filename = get_output_filename()
+makedirs(output_dirname, exist_ok=True) # Create the output directory if not exists
+
+with open(path.join(output_dirname, "school-center-distance.tsv"), 'w', encoding='utf-8') as intermediate_file, \
+open(path.join(output_dirname, 'output_filename'), 'w', encoding='utf-8') as a_file:
```
Comment (sapradhan) [inline school_center.py:218]: should this be just output_filename, not in quotes ?

Comment (sebakthapa) [reply school_center.py]: Yes, Thank you, that was my mistake. I corrected it.

**Comment by sebakthapa** [comment]
@sumanashrestha , changes are done to meet all the requirements you mentioned in the [PR discussion](https://github.com/moest-np/center-randomize/pull/54#issuecomment-2075532201).

For writing test, I think, it would be better to write modular tests but right now main logic is executed outside of any function.
This PR #65  addresses in making main code modular. 

However, for now it is **properly tested** manually and **is ready to merge**.

File: school_center.py [hunk 2]
```diff
@@ -231,10 +224,30 @@ def is_allocated(scode1: str, scode2: str) -> bool:
...
+        return dirname
+    else:
+        return DEFAULT_OUTPUT_DIR
+
+def get_output_filename():
+    basename = path.basename(args.output)
+    if(basename):
+        return basename
+    else:
+        return DEFAULT_OUTOUT_FILENAME
```
Comment (sapradhan) [inline school_center.py:240]: typo DEFAULT_OUTPUT_FILENAME

File: school_center.py [hunk 3]
```diff
@@ -231,10 +224,30 @@ def is_allocated(scode1: str, scode2: str) -> bool:
...
+
+
+output_dirname = get_output_dir()
+output_filename = get_output_filename()
+makedirs(output_dirname, exist_ok=True) # Create the output directory if not exists
+
+print(path.join(output_dirname, output_filename))
```
Comment (sapradhan) [inline school_center.py:247]: either remove or use logger.info

Comment (sebakthapa) [reply school_center.py]: Thank you. It has been resolve now.

**Pull request merged**